src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -751,6 +741,117 @@ private void ReturnReadBufferIfEmpty()
                 _decryptedBytesCount = 0;
                 _decryptedBytesOffset = 0;
             }
+            else if (_decryptedBytesCount == 0)
+            {
+                _decryptedBytesOffset = 0;
+            }
+        }
+
+        private bool haveFullTlsFrame(ref int frameSize)
+        {
+            if (_internalBufferCount < SecureChannel.ReadHeaderSize)
+            {
+                frameSize = int.MaxValue;
+                return false;
+            }
+
+            frameSize = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
+            return _internalBufferCount >= frameSize;
+        }
+
+        private ValueTask<int> getFullFrameIfNeed<TIOAdapter>(TIOAdapter adapter)
+            where TIOAdapter : IReadWriteAdapter
+        {
+            int frameSize;
+            if (_internalBufferCount >= SecureChannel.ReadHeaderSize)
+            {
+                frameSize = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
+                if (_internalBufferCount >= frameSize)
+                {
+                    return new ValueTask<int>(frameSize);
+                }
+            }
+            else
+            {
+                frameSize = int.MaxValue;
+            }
+
+            // We may have enough space to complete frame, but we may still do extra IO if the frame is small.
+            // So we will attempt larger read - that is trade of with extra copy.
+            // This may be updated at some point based on size of existing chunk, rented buffer and size of 'buffer'.
+            ResetReadBuffer();
+
+            // _internalOffset is 0 after ResetReadBuffer and we use _internalBufferCount to determined where to read.
+            while (_internalBufferCount < frameSize)
+            {
+                // We don't have enough bytes buffered, so issue an initial read to try to get enough.  This is
```

Comment:
This comment isn't really correct anymore

Reply:
What do you think is incorrect? I'm reading it again and it seems to describe the intend.